From 5454a7e854b9ecb6f37ea17599c95cbbdb7fcdb0 Mon Sep 17 00:00:00 2001
From: Giovanni Mascellani <gmascellani@codeweavers.com>
Date: Fri, 17 Jan 2025 14:45:55 +0100
Subject: [PATCH] dxgi: Directly signal the frame latency fence.

Instead of going through the command queue again. That's because at
that point we just submitted the frame for presentation to Vulkan,
so the best approximation we have for when the frame will be
presented is now, not after the currently outstanding work in the
command queue is processed. It's still a rather poor approximation,
but for something better we need VK_KHR_present_wait.
---
 dlls/dxgi/swapchain.c  | 3 +--
 dlls/dxgi/tests/dxgi.c | 1 -
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/dlls/dxgi/swapchain.c b/dlls/dxgi/swapchain.c
index 35460691918..94551fe14eb 100644
--- a/dlls/dxgi/swapchain.c
+++ b/dlls/dxgi/swapchain.c
@@ -2296,8 +2296,7 @@ static HRESULT d3d12_swapchain_op_present_execute(struct d3d12_swapchain *swapch
 
     if (swapchain->frame_latency_fence)
     {
-        if (FAILED(hr = ID3D12CommandQueue_Signal(swapchain->command_queue,
-                swapchain->frame_latency_fence, op->present.frame_number + 1)))
+        if (FAILED(hr = ID3D12Fence_Signal(swapchain->frame_latency_fence, op->present.frame_number + 1)))
         {
             ERR("Failed to signal frame latency fence, hr %#lx.\n", hr);
             return hr;
diff --git a/dlls/dxgi/tests/dxgi.c b/dlls/dxgi/tests/dxgi.c
index 4ff27715459..b8cf3597483 100644
--- a/dlls/dxgi/tests/dxgi.c
+++ b/dlls/dxgi/tests/dxgi.c
@@ -7222,7 +7222,6 @@ static void test_frame_latency_event(IUnknown *device, BOOL is_d3d12)
         for (i = 0; i < 3; i++)
         {
             wait_result = WaitForSingleObject(semaphore, 100);
-            todo_wine
             ok(!wait_result, "Got unexpected wait result %#lx.\n", wait_result);
         }
 
-- 
GitLab

